AP_RangeFinder: ensure that too low and too far states are propagated in DroneCAN

andyp1per opened the pull request:
The current code means that you will always get "No Data" (which is fatal) rather than too low (which happens when you are sitting on the ground).

tridge commented on libraries/AP_RangeFinder/AP_RangeFinder_DroneCAN.cpp:
```diff
@@ -81,7 +81,14 @@ AP_RangeFinder_DroneCAN* AP_RangeFinder_DroneCAN::get_dronecan_backend(AP_DroneC
+    if (_status != RangeFinder::Status::Good) {
+        //handle additional states received by measurement handler
+        if (_status == RangeFinder::Status::OutOfRangeLow) {
+            state.distance_m = 0;
```
please don't force zero, the actual value is useful in logs

andyp1per replied:
Well all the bear traps are in readRangeFinder() as you know. If we really want this for logging I think we should have a last_distance_read() and return nan() if it never read a distance and then use that for logging.

rmackay9 commented:
It's an existing issue so not a blocker but I think it would be good to put comments in the AP_RangeFinder_DroneCAN.h file beside the internal variables (e.g. _status) to clarify that these are the values received from the sensor.  maybe it's just me but I find it quite confusing to have set_status(_status).

andyp1per commented on libraries/AP_RangeFinder/AP_RangeFinder_DroneCAN.cpp:
```diff
@@ -81,7 +81,14 @@ AP_RangeFinder_DroneCAN* AP_RangeFinder_DroneCAN::get_dronecan_backend(AP_DroneC
 void AP_RangeFinder_DroneCAN::update()
 {
     WITH_SEMAPHORE(_sem);
-    if ((AP_HAL::millis() - _last_reading_ms) > 500) {
```
Fixed

andyp1per commented:
> I think that @peterbarker also suggested that we change the logic in the update() method to have early returns and/or a switch statement.

Fixed
 
> I think the logic could be:
> 
>     * always update the distance with whatever is received from the sensor

Yes, but in this case the distance we have received is "TooLow" the only thing we can then usefully set it to is 0

> 
> 
> Re the status:
> 
>     1. if we haven't received an update in 500ms (or whatever) set_status to NoData
> 
>     2. if the status from the sensor is NoData then we set status to NoData (e.g. the sensor timed out)
> 
>     3. if the status from the sensor is TooLow or TooHigh we set status to the sensor's status

Fixed these

> 
>     4. if the status from the sensors is Good then we compare the distance to the min/max parameters and if it's out-of-range then we set status to TooLow or TooHigh appropriately.  otherwise we set status to Good.

This was already happening

peterbarker commented on libraries/AP_RangeFinder/AP_RangeFinder_DroneCAN.cpp:
```diff
@@ -82,17 +82,24 @@ void AP_RangeFinder_DroneCAN::update()
+        return;
+    }
+
+    if (_status == RangeFinder::Status::Good && new_data) {
```
```suggestion
    if (!new_data) {
        return;
    }
```

(the check for `Good` is redundant with the block just above it)

tridge commented:
@andyp1per should set new_data on all types of new data from device

andyp1per commented on libraries/AP_RangeFinder/AP_RangeFinder_DroneCAN.cpp:
```diff
@@ -81,19 +81,28 @@ AP_RangeFinder_DroneCAN* AP_RangeFinder_DroneCAN::get_dronecan_backend(AP_DroneC
         set_status(_status);
     }
+
+    new_data = false;
```
No we do not. We absolutely want too low to have a chance if registering 0 if its on the ground otherwise everything gets confused. I already took out the code to set explicitly to 0 because everyone want use to just us the last value. I'm not going to change it to some different wrong value.

peterbarker reviewed (approved):
LGTM.  Can be merged after testing
